Playwright: migrate WordPress.com Free/Publish subsuite of Signup spec.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes to combine and migrate subsuites in the Selenium `Signup` spec:
- `Sign up for a free WordPress.com site from the Jetpack new site page, and log in via a magic link`
- `Sign up for a free site, see the site preview, activate email and can publish`

Key changes:
- combine and streamline the subsuites from Selenium `Signup` spec.
- add ability to include subject filter in EmailClient.
- add new POMs and components for interaction with the `/me` endpoint.
- add ability in GutenbergEditorPage to interact with the sidebar.
- add ability for LoginPage to interact with magic links.

#### Testing instructions

Locally
- pull branch
- open `test/e2e/specs/specs-playwright/wp-signup__free-publish.ts`
- remove `.skip` annotation at top level suite.
- run suite with `yarn jest --detectOpenHandles specs/specs-playwright/wp-signup__free-publish.ts` for desktop viewport.

Related to #

worldomonation committed:
Migrate Invite: Signup/Remove spec, similar to Invite: Viewer/Editor spec in Selenium.
- simplify steps where appropriate.
- skip test by default for now.

worldomonation committed:
Rename spec file name and suite title.
Remove unnecessary `throw new Error`.

worldomonation committed:
Migrate Signup: WordPress.com Free/Publish subspec from `wp-signup-spec.js`.
- new methods and POMs.
- reorganize flow to test the signup, WordPress.com Free plan, publish, magic link login then delete site.

worldomonation committed:
Fixes and changes:
- EmailClient now takes additional optional parameter `subject`.
- implement POMS and objects for /me endpoint.
- remove unnecessary wait on GutenbergEditorPage.toggleSidebar.
- correct test steps.
- fix spec for mobile viewport.

worldomonation committed:
Skip by default.
Fix index.ts.

worldomonation committed:
Remove commented out code.
Adjust timeout in selectPlan.

worldomonation commented on test/e2e/specs/specs-playwright/wp-signup__free-publish.ts:
```diff
@@ -0,0 +1,138 @@
+} from '@automattic/calypso-e2e';
+import { Page } from 'playwright';
+
+describe.skip( DataHelper.createSuiteTitle( 'Signup: WordPress.com Free/Publish' ), function () {
```
Skipped by default as this shouldn't need to be run every commit.

The intention is for this test to be run only for trunk merges (pre-release).

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -61,4 +81,51 @@ export class LoginPage {
+	 * Navigates to the Login page.
+	 */
+	async visit(): Promise< void > {
+		await this.page.goto( getCalypsoURL( 'log-in' ), { waitUntil: 'networkidle' } );
```
Normally, the navigation part is handled in `LoginFlow` as part of the different login options, but I am leaning towards simplifying the structure of LoginPage and LoginFlow by merging the two.

It will be something handled in a follow-up PR, as this one is big enough already.

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -35,6 +48,18 @@ export class LoginPage {
+	/**
+	 * Switch account if an user is already logged in.
+	 */
+	private async switchAccount(): Promise< void > {
```
Extracted from `login`.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -302,4 +307,20 @@ export class GutenbergEditorPage {
+	/**
+	 * Opens the Block Editor sidebar (left hand side).
+	 */
+	async toggleSidebar(): Promise< void > {
```
Toggles the sidebar by clicking on the `W`.

Note, on mobile this cannot be used unless in landscape mode - see https://github.com/Automattic/wp-calypso/issues/51162.

worldomonation committed:
Rename NavbarComponent.clickAccountSettings.
Rename spec file.

dpasque commented on packages/calypso-e2e/src/lib/components/domain-search-component.ts:
```diff
@@ -0,0 +1,66 @@
+
+const selectors = {
+	searchInput: `input[aria-label="What would you like your domain name to be?"]`,
+	placeholder: `.is-placeholder`,
```
Nit: it took me a second to discover what placeholder this was (I was wondering if it was the search box one). What do you think of `resultPlaceholder` as a more clarified name?

worldomonation replied:
Good suggestion, fixed in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on packages/calypso-e2e/src/lib/components/domain-search-component.ts:
```diff
@@ -0,0 +1,66 @@
+	 * @returns {string} Domain that was selected.
+	 */
+	async selectDomain( keyword: string ): Promise< string > {
+		const selector = `${ selectors.resultItem }:has-text("${ keyword }")`;
```
Nit suggestion: want to just make this a dynamic selector in the constants above? We only seem to care about result items based on keyword

worldomonation replied:
Ah, makes sense. Fixed in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on packages/calypso-e2e/src/lib/components/me/me-sidebar-component.ts:
```diff
@@ -0,0 +1,29 @@
+	async navigate( menu: string ): Promise< void > {
+		await Promise.all( [
+			this.page.waitForNavigation(),
+			this.page.click( `a:has(span:has-text("${ menu }") )` ),
```
Optional suggestion: what do you think trying to lock this selector down to the sidebar by including a parent element of the sidebar? I wonder if "a link with the text in it" might be just a little risky for collisions on page content.

worldomonation replied:
Moved selector up to an object and made it dynamic, pinned with `.sidebar` as the prefix in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on packages/calypso-e2e/src/lib/pages/account-closed-page.ts:
```diff
@@ -0,0 +1,30 @@
+
+const selectors = {
+	accountClosedMessage: ':text("Your account has been closed")',
+	accountClosedSubmessage: ':text("Thanks for flying with WordPress.com")',
```
Optional: do you think we need this submessage? This feels like something quirky we might tweak semi-regularly, and I wonder if it's enough to just confirm the main closure message.

worldomonation replied:
Hmm, you're right. Let's rely on less selectors.

Removed in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -327,4 +332,20 @@ export class GutenbergEditorPage {
+	/**
+	 * Opens the Block Editor sidebar (left hand side).
+	 */
+	async toggleSidebar(): Promise< void > {
```
Can we give this method a slightly more descriptive name to differentiate it from the settings sidebar? Just that sidebar is an overloaded term with the settings sidebar, and I know sometimes folks just go off of method name when working quickly.

worldomonation replied:
Yeah, you're right - the DOM references this element as `navsidebar`, so how about that?

worldomonation replied:
Updated to `openNavSidebar` in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -327,4 +332,20 @@ export class GutenbergEditorPage {
+	/**
+	 * Opens the Block Editor sidebar (left hand side).
+	 */
+	async toggleSidebar(): Promise< void > {
```
Do we want to make this match the behavior of the settings sidebar where instead of just toggling, we have methods for opening and closing that only take action if the sidebar isn't already opened or closed (respectively?)

Toggling to me always feels a little risky, as you may not know the starting state!

worldomonation replied:
Will change to just `openSidebar` for now as no tests need to close it yet - we'll make it when that time comes.

Fixed in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -61,4 +81,51 @@ export class LoginPage {
+	/**
+	 * Clicks on a link to navigate to the account signup page.
+	 */
+	async clickSignup(): Promise< void > {
```
When I first read the method name and header, I was surprised then by all the navigation that this method did. What do you think of renaming this to capture the fact that it does all the navigation and clicks the signup button. Something like `navigateAndStartSignup`?

worldomonation replied:
I can see how it is a bit awkward, though it's partially due to the in-between state of the `LoginPage`. I do plan on a large refactor to eliminate the use of either `LoginFlow` or `LoginPage`, because having both feels rather unnecessary now (and is a legacy of the way Selenium suite is set up). 

I would expect the methods and names to be shuffled around when that time comes. Would this be acceptable until then?

And yes, I will definitely refactor this - not am empty promise 😛

dpasque replied:
Of course, that sounds great! :)

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__free.ts:
```diff
@@ -0,0 +1,138 @@
+} from '@automattic/calypso-e2e';
+import { Page } from 'playwright';
+
+describe.skip( DataHelper.createSuiteTitle( 'Signup: WordPress.com Free' ), function () {
```
Nit: should we reflect in the suite title that this flow also includes coverage of Magic Links?

worldomonation replied:
Perhaps. I will include features touched other than signup in the suite name.

worldomonation replied:
Fixed in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__free.ts:
```diff
@@ -0,0 +1,138 @@
+} from '@automattic/calypso-e2e';
+import { Page } from 'playwright';
+
+describe.skip( DataHelper.createSuiteTitle( 'Signup: WordPress.com Free' ), function () {
```
Now that we have `jest-runner-groups`, you could remove this `skip()` and instead just include this in the currently unused `calypso-release` group:

```typescript
/**
* @group calypso-release
* /
```

worldomonation replied:
Nice! Added bonus, VSCodium won't be annoying me with a wall of yellow squiggly lines.

worldomonation replied:
Fixed in bd8bc48bb555cb23660d753d3d43076801065851.

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__free.ts:
```diff
@@ -0,0 +1,138 @@
+		it( 'Return to Home dashboard', async function () {
+			// Temporary workaround due to https://github.com/Automattic/wp-calypso/issues/51162.
+			// Conditional can be removed once fixed.
+			if ( BrowserHelper.getViewportName() === 'mobile' ) {
```
This method name changed - should now be `getTargetDeviceName()`

worldomonation replied:
Ahh yes, thanks!

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__free.ts:
```diff
@@ -0,0 +1,138 @@
+		} );
+	} );
+
+	describe( 'Interact', function () {
```
Hypothetical question here: do you think it's enough to just make sure you can launch the editor? We have a lot of other post flow creation covered elsewhere. TBH, I'm not sure if publishing right after signup has broken specifically before, but I think making sure the editor loads and then backing out is enough here, and will help keep the spec leaner!

worldomonation replied:
True, I think this might be overtesting for no reason. Keen eye, thanks for that!

worldomonation replied:
Removed publish post part of the test in bd8bc48bb555cb23660d753d3d43076801065851. Also renamed to `Interact with editor`.

dpasque reviewed (commented):
@worldomonation Left a bunch of questions/nits/optional comments - feel free to just do the ones that you think are helpful!

The only important comment is the changed device method name in `BrowserHelper`!

Also, you did a really nice job with the flow creation here. Everything reads so linearly and clearly - it's like a perfect showcase of Playwright's strengths! Nice work! 😄

worldomonation committed:
Address PR review comments.
- fixed method name for `getTargetDevice`.
- more descriptive selector name for placeholder.
- fix selectors in MeSidebarComponent and DomainSearchComponent.
- remove submessage from AccountClosedPage.
- rename toggleSidebar to openNavSidebar.
- add jest runner group header.

dpasque commented on packages/calypso-e2e/src/lib/components/domain-search-component.ts:
```diff
@@ -0,0 +1,70 @@
+const selectors = {
+	searchInput: `input[aria-label="What would you like your domain name to be?"]`,
+	resultPlaceholder: `.is-placeholder`,
+	resultItem: ( keyword: string ) => `.domain-suggestion__content:has-text(${ keyword })`,
```
Ah, I think you're missing quotes around here! It fails otherwise unfortunately!
```suggestion
	resultItem: ( keyword: string ) => `.domain-suggestion__content:has-text("${ keyword }")`,
```

worldomonation replied:
Good catch, thanks. Not sure how that made it through.

worldomonation replied:
Fixed in 3c5d75c615396095f70524b6e99748c0ad17f321.

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__free.ts:
```diff
@@ -0,0 +1,137 @@
+				if ( BrowserHelper.getTargetDeviceName() === 'mobile' ) {
+					await page.goBack();
+				} else {
+					await gutenbergEditorPage.openNavSidebar();
```
Looks like you're missing the class instantiation here!
```suggestion
					gutenbergEditorPage = new GutenbergEditorPage( page );
					await gutenbergEditorPage.openNavSidebar();
```

worldomonation replied:
Hmm, rookie mistake, thanks for catching it. Fixed in 3c5d75c615396095f70524b6e99748c0ad17f321.

dpasque reviewed (approved):
The updates you made look great!

It looks like there's a couple of runtime errors during the script execution, I've added suggestion comments that should resolve them!

Otherwise, this looks gtg, so approving so you can merge after those two comments! Nice work 😄

worldomonation committed:
Address PR comments.
- fix selector quotes.
- instantiate GutenbergEditorComponent in the spec before using it.
